contracts/DomainRegistryImplementationV2.sol

```diff
@@ -0,0 +1,126 @@
+        bool isRegistered;
+        address owner;
+    }
+
+    mapping(string => Domain) public domains;
+    mapping(string => uint256) public rewards;
+    mapping(string => bool) public rewardsClaimed;
+    mapping(string => uint256) public domainRewards;
+
+    event DomainRegistered(string indexed domainName, address indexed owner);
+    event DomainReleased(string indexed domainName);
+    event RewardSet(string indexed domainName, uint256 rewardAmount);
+    event RewardClaimed(string indexed childDomain, uint256 rewardAmount);
+
+function initialize() public initializer {
+    __Ownable_init();
+    domainPrice = 1 ether; 
+}
+
+    modifier isValidDomain(string memory _domainName) {
+        _domainName = _domainName.stripProtocol();
+        require(bytes(_domainName).length > 0, "Domain is empty");
+        bytes memory domainBytes = bytes(_domainName);
+        require(domainBytes[domainBytes.length - 1] != bytes1('.'), "Domain ends with a dot");
+        _;
+    }
+
+    modifier domainOwnedBySender(string memory domainName) {
+        domainName = domainName.stripProtocol();
+        require(domains[domainName].owner == msg.sender, "Not the domain owner");
+        _;
+    }
+
+    modifier domainExistence(string memory domainName, bool shouldExist) {
+        bool exists = domains[domainName.stripProtocol()].owner != address(0);
+        if (shouldExist) {
+            require(exists, "Domain doesn't exist");
+        } else {
+            require(!exists, "Domain exists");
+        }
+        _;
+    }
+
+   function setDomainPrice(uint256 _price) external onlyOwner {
+        domainPrice = _price;
+    }
+
+    function setRewardForChildDomains(string memory domainName, uint256 rewardAmount) external payable domainExistence(domainName, true) domainOwnedBySender(domainName) {
+        require(msg.value == rewardAmount, "Incorrect amount sent");
+        rewards[domainName] = rewardAmount;
+        emit RewardSet(domainName, rewardAmount);
+    }
+
+function setDomainReward(string memory domainName, uint256 reward) external onlyOwner {
+    // Moved onlyOwner check to the start
+    require(domains[domainName].isRegistered, "Domain doesn't exist");
+    domainRewards[domainName] = reward;
+}
+
+    function claimReward(string memory childDomain) external domainExistence(childDomain, true) domainOwnedBySender(childDomain) {
+        require(!rewardsClaimed[childDomain], "Reward already claimed");
+        string memory parentDomain = DomainUtils.extractParentDomain(childDomain);
+        require(rewards[parentDomain] > 0, "No reward set by parent domain");
+        
+        uint256 rewardAmount = rewards[parentDomain];
+        rewards[parentDomain] = 0; // Reset reward amount for parent domain
+        rewardsClaimed[childDomain] = true; // Mark reward as claimed for child domain
+
+        payable(domains[childDomain].owner).sendValue(rewardAmount);
+        emit RewardClaimed(childDomain, rewardAmount);
+    }
```

Comment:
Why does the domain controller need to claim reward for each child domain instead to claim all accumulated reward that belongs to him/her (over all domains)? It will simplify a lot.
 -5 points